python/ray/tune/schedulers/resource_changing_scheduler.py

```diff
@@ -14,130 +16,374 @@
+        )
+        return total_available_cpus, total_available_gpus
 
-        # Set upper limits for resources based on number of live trials
-        # to ensure that the trial cannot get more resources that it's
-        # possible to run
-        num_running_trials = len(trial_runner.get_live_trials())
-        if min_cpu == 0:
+    def _get_upper_limits(self) -> Tuple[int, int]:
+        num_running_trials = len(self.trial_runner.get_live_trials())
+        if self.min_cpu == 0:
             upper_cpu_limit = 0
         else:
-            upper_cpu_limit = math.ceil(total_available_cpus / num_running_trials)
+            upper_cpu_limit = math.ceil(self.total_available_cpus / num_running_trials)
             # Round to nearest bundle minimum
             # eg. 8 CPUs between 3 trials with min 2 CPUs per bundle
             #   -> 4, 2, 2
             if self.add_bundles:
                 upper_cpu_limit = (
-                    math.ceil(upper_cpu_limit / min_cpu_bundle) * min_cpu_bundle
+                    math.ceil(upper_cpu_limit / self.min_cpu_bundle)
+                    * self.min_cpu_bundle
                 )
-            upper_cpu_limit = max(min_cpu, upper_cpu_limit)
+            upper_cpu_limit = max(self.min_cpu, upper_cpu_limit)
 
-        if min_gpu == 0:
+        if self.min_gpu == 0:
             upper_gpu_limit = 0
         else:
-            upper_gpu_limit = math.ceil(total_available_gpus / num_running_trials)
+            upper_gpu_limit = math.ceil(self.total_available_gpus / num_running_trials)
             # Ensure we don't go below per-bundle minimum
             if self.add_bundles:
                 upper_gpu_limit = (
-                    math.ceil(upper_gpu_limit / min_cpu_bundle) * min_gpu_bundle
+                    math.ceil(upper_gpu_limit / self.min_gpu_bundle)
+                    * self.min_gpu_bundle
                 )
-            upper_gpu_limit = max(min_gpu, upper_gpu_limit)
+            upper_gpu_limit = max(self.min_gpu, upper_gpu_limit)
 
-        # Function to check how many CPUs and GPUs a trial is using currently
-        def get_used_cpus_and_gpus(t: Trial):
-            return (
-                t.placement_group_factory.required_resources.get("CPU", 0),
-                t.placement_group_factory.required_resources.get("GPU", 0),
+        return upper_cpu_limit, upper_gpu_limit
+
+    def _modify_upper_limits_with_increase_by_times(self) -> None:
+        if self.increase_by and self.increase_by_times > 0:
+            required_cpus = self.increase_by.get("CPU", 0)
+            required_gpus = self.increase_by.get("GPU", 0)
+            self.upper_cpu_limit = min(
+                self.upper_cpu_limit,
+                self.min_cpu + required_cpus * self.increase_by_times,
+            )
+            self.upper_gpu_limit = min(
+                self.upper_gpu_limit,
+                self.min_gpu + required_gpus * self.increase_by_times,
             )
```

Comment:
I think this is related to my general question about this Callable class. Do we need these to be class var? 
In other words, I find the situation confusing that function A (not `__init__`) populates `self.a`, which is then used in function B under the assumption that function B always happens after function A. And as a reader I have to keep track of all that.

Reply:
All those variables are instantiated from scratch every time the callable is called. They are made attributes, so that they can be easily accessed in the methods without always passing a dozen or so arguments.

Reply:
Sure, if you prefer it this way, then can I ask that the function name to be more explicit, like `def set_a()` - then we all know what it is doing without looking into its implementation. Also it may be helpful to be consistent. If we choose the way of `_modify_upper_limits_with_increase_by_times`, we may want to consider changing `self.upper_cpu_limit, self.upper_gpu_limit = self._get_upper_limits()` to just `self._set_upper_limits()`. The same goes for `get_used_cpus_and_gpus `

Reply:
Ok, let me change all of those from attributes to normal variables. That should make it clearer.